daemon.c

```diff
@@ -734,22 +758,66 @@ static void register_mount_service(SVCXPRT * udptransp, SVCXPRT * tcptransp)
 static SVCXPRT *create_udp_transport(unsigned int port)
 {
     SVCXPRT *transp = NULL;
-    struct sockaddr_in sin;
     int sock;
-    const int on = 1;
 
-    /* Make sure we null the entire sockaddr_in structure */
-    memset(&sin, 0, sizeof(struct sockaddr_in));
+    sock = socket(PF_INET6, SOCK_DGRAM, 0);
 
-    if (port == 0)
-	sock = RPC_ANYSOCK;
-    else {
-	sin.sin_family = AF_INET;
-	sin.sin_port = htons(port);
-	sin.sin_addr.s_addr = opt_bind_addr.s_addr;
+    if ((sock == -1) && (errno == EAFNOSUPPORT))
 	sock = socket(PF_INET, SOCK_DGRAM, 0);
+
+    if (sock == -1) {
+	perror("socket");
+	fprintf(stderr, "Couldn't create a listening udp socket\n");
+	exit(1);
+    }
+
+    if (port != 0) {
+	int domain;
+	socklen_t len;
+	const int on = 1;
+
+	const struct sockaddr *sin;
+	size_t sin_len;
+	struct sockaddr_in sin4;
+	struct sockaddr_in6 sin6;
+
+	len = sizeof(domain);
+	if (getsockopt(sock, SOL_SOCKET, SO_DOMAIN, &domain, &len)) {
+	    perror("getsockopt");
+	    fprintf(stderr, "Couldn't create a listening udp socket\n");
+	    exit(1);
+	}
+
+	if (domain == PF_INET6) {
+	    /* Make sure we null the entire sockaddr_in6 structure */
+	    memset(&sin6, 0, sizeof(struct sockaddr_in6));
+
+	    sin6.sin6_family = AF_INET6;
+	    sin6.sin6_port = htons(port);
+	    sin6.sin6_addr = opt_bind_addr;
+
+	    sin = (const struct sockaddr*)&sin6;
+	    sin_len = sizeof(sin6);
+	} else {
+	    if (!IN6_IS_ADDR_V4MAPPED(&opt_bind_addr) &&
+                !IN6_IS_ADDR_V4COMPAT(&opt_bind_addr)) {
```

Comment:
The default value of `opt_bind_addr` (address `::`) is **false** for both [IN6_IS_ADDR_V4MAPPED](https://code.woboq.org/userspace/glibc/inet/netinet/in.h.html#436) and [IN6_IS_ADDR_V4COMPAT](https://code.woboq.org/userspace/glibc/inet/netinet/in.h.html#443). Allow the address by adding check for [IN6_IS_ADDR_UNSPECIFIED](https://code.woboq.org/userspace/glibc/inet/netinet/in.h.html#460) which will match only this one address. It can then be converted to IPv4 as `0.0.0.0`.

Reply:
Ah, thanks. Should be fixed now.